File: roboplan_ros_visualization/src/interactive_marker_ik.cpp
#include "roboplan_ros_visualization/interactive_marker_ik.hpp"

namespace roboplan_ros_visualization {

InteractiveMarkerIK::InteractiveMarkerIK(rclcpp::Node::SharedPtr node,
                                         std::shared_ptr<roboplan::Scene> scene,
                                         const std::string& joint_group,
                                         const std::string& base_link, const std::string& tip_link,
                                         const roboplan::SimpleIkOptions& options,
                                         const std::string& namespace_name)
    : node_(node), scene_(scene), namespace_(namespace_name), joint_group_(joint_group) {

  // Get joint group information
  auto joint_group_info = scene_->getJointGroupInfo(joint_group_);
  if (!joint_group_info.has_value()) {
    throw std::runtime_error("Failed to get joint group info for: " + joint_group_);
  }

  // Construct the IK solver
  ik_solver_ = std::make_shared<roboplan_ros_cpp::RoboPlanIK>(scene_, joint_group_, base_link,
                                                              tip_link, options);

  // Set initial states based on the state of the scene
  last_joint_positions_ = scene_->getCurrentJointPositions();
  Eigen::Matrix4d se3_pose = scene_->forwardKinematics(last_joint_positions_, tip_link);
  current_pose_ = roboplan_ros_cpp::se3ToPose(se3_pose);

  // Create joint state publisher
  std::string joint_state_topic = namespace_name + "/joint_states";
  joint_state_pub_ = node_->create_publisher<sensor_msgs::msg::JointState>(joint_state_topic, 1);

  // Create the interactive marker server
  ik_server_ = std::make_shared<interactive_markers::InteractiveMarkerServer>(namespace_, node_);
  createInteractiveMarker(current_pose_);
  ik_server_->applyChanges();

  // Solve and publish joint states at 20 Hz
  joint_state_timer_ =
      node_->create_wall_timer(std::chrono::milliseconds(50), [this]() { solveIK(); });

  RCLCPP_INFO(node_->get_logger(), "Constructed interactive marker controller");
  RCLCPP_INFO(node_->get_logger(), "Joint group: %s", joint_group_.c_str());
  RCLCPP_INFO(node_->get_logger(), "Base link: %s", base_link.c_str());
  RCLCPP_INFO(node_->get_logger(), "End-effector: %s", tip_link.c_str());
}

void InteractiveMarkerIK::createInteractiveMarker(const geometry_msgs::msg::Pose& pose) {
  visualization_msgs::msg::InteractiveMarker int_marker;
  int_marker.header.frame_id = namespace_ + "/" + ik_solver_->getBaseFrame();
  int_marker.name = "ik_target";
  int_marker.description = "IK Target Pose for " + joint_group_;
  int_marker.pose = pose;
  int_marker.scale = 0.2;

  visualization_msgs::msg::InteractiveMarkerControl control;

  control.orientation.w = 1.0;
  control.orientation.x = 1.0;
  control.orientation.y = 0.0;
  control.orientation.z = 0.0;
  control.name = "move_x";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::MOVE_AXIS;
  int_marker.controls.push_back(control);

  control.orientation.w = 1.0;
  control.orientation.x = 0.0;
  control.orientation.y = 1.0;
  control.orientation.z = 0.0;
  control.name = "move_y";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::MOVE_AXIS;
  int_marker.controls.push_back(control);

  control.orientation.w = 1.0;
  control.orientation.x = 0.0;
  control.orientation.y = 0.0;
  control.orientation.z = 1.0;
  control.name = "move_z";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::MOVE_AXIS;
  int_marker.controls.push_back(control);

  control.orientation.w = 1.0;
  control.orientation.x = 1.0;
  control.orientation.y = 0.0;
  control.orientation.z = 0.0;
  control.name = "rotate_x";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::ROTATE_AXIS;
  int_marker.controls.push_back(control);

  control.orientation.w = 1.0;
  control.orientation.x = 0.0;
  control.orientation.y = 1.0;
  control.orientation.z = 0.0;
  control.name = "rotate_y";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::ROTATE_AXIS;
  int_marker.controls.push_back(control);

  control.orientation.w = 1.0;
  control.orientation.x = 0.0;
  control.orientation.y = 0.0;
  control.orientation.z = 1.0;
  control.name = "rotate_z";
  control.interaction_mode = visualization_msgs::msg::InteractiveMarkerControl::ROTATE_AXIS;
  int_marker.controls.push_back(control);

  // Add the marker to the server with the feedback callback
  ik_server_->insert(int_marker, std::bind(&InteractiveMarkerIK::markerFeedbackCallback, this,
                                           std::placeholders::_1));
}

void InteractiveMarkerIK::markerFeedbackCallback(
    const visualization_msgs::msg::InteractiveMarkerFeedback::ConstSharedPtr& feedback) {
  // On pose updates set the latest pose and let the timer manage solving for joint states
  if (feedback->event_type == visualization_msgs::msg::InteractiveMarkerFeedback::POSE_UPDATE) {
    std::lock_guard<std::mutex> lock(state_mutex_);
    current_pose_ = feedback->pose;
  }
}

void InteractiveMarkerIK::solveIK() {
  auto joint_positions = ik_solver_->solveIK(current_pose_, last_joint_positions_);

  if (joint_positions.has_value()) {
    last_joint_positions_ = joint_positions.value();
    publishJointStates();
  } else {
    RCLCPP_WARN_THROTTLE(node_->get_logger(), *node_->get_clock(), 1000,
                         "IK failed to find solution for target pose");
  }
}

void InteractiveMarkerIK::publishJointStates() {
  auto msg = sensor_msgs::msg::JointState();
  msg.header.stamp = node_->now();
  msg.header.frame_id = "";
  msg.name = scene_->getJointNames();

  {
    // Convert Eigen to std::vector
    std::lock_guard<std::mutex> lock(state_mutex_);
    msg.position.resize(last_joint_positions_.size());
    for (int i = 0; i < last_joint_positions_.size(); ++i) {
      msg.position[i] = last_joint_positions_(i);
    }
  }

  joint_state_pub_->publish(msg);
}

void InteractiveMarkerIK::updateMarkerPose(const geometry_msgs::msg::Pose& pose) {
  ik_server_->setPose("ik_target", pose);
  ik_server_->applyChanges();
  current_pose_ = pose;
}

}  // namespace roboplan_ros_visualization

File: roboplan_ros_visualization/roboplan_ros_visualization/interactive_marker_ik.py
#!/usr/bin/env python3

import os
import threading
from collections.abc import Callable
from typing import Optional
import numpy as np

from rclpy.node import Node
from rclpy.executors import SingleThreadedExecutor, ExternalShutdownException
from rclpy.qos import (
    QoSProfile,
    QoSReliabilityPolicy,
    QoSHistoryPolicy,
    QoSDurabilityPolicy,
)
from geometry_msgs.msg import Pose
from sensor_msgs.msg import JointState
from interactive_markers import InteractiveMarkerServer
from visualization_msgs.msg import (
    InteractiveMarker,
    InteractiveMarkerControl,
    InteractiveMarkerFeedback,
    Marker,
)
from roboplan.core import Scene
from roboplan.simple_ik import SimpleIkOptions
from roboplan_ros_py.kinematics import RoboPlanIK
from roboplan_ros_cpp.bindings import poseToSE3, se3ToPose


class InteractiveMarkerIK:
    """
    Utility for creating an interactive marker in RViz and passing published
    poses to an IK solver. Useful for nodes that ingest poses from users and
    publish or process IK solutions for those poses.
    """

    def __init__(
        self,
        node: Node,
        scene: Scene,
        joint_group: str,
        base_link: str,
        tip_link: str,
        options: SimpleIkOptions = SimpleIkOptions(),
        solve_callback: Optional[Callable] = None,
        node_name: str = "imarker_ik_node",
        namespace: str = "/roboplan_ik",
    ):
        """
        Initialize the interactive marker IK controller.

        Args:
            node: ROS node instance to be passed to the InteractiveMarker server.
            scene: A fully configured RoboPlan scene.
            joint_group: The joint group for the IK solver
            base_link: Base link of the IK chain.
            tip_link: Tip link of the iK chain.
            options: Options for the IK solver.
            solve_callback: Callback function to immediately handle IK solutions.
            node_name: Name for the dedicated Node for callbacks to the IK marker.
            namespace: Namespace for the InteractiveMarkerServer and JointStatePublisher.
        """
        self.node = node
        self.scene = scene
        self.joint_group = joint_group
        self.base_link = base_link
        self.tip_link = tip_link
        self.namespace = namespace
        self.solve_callback = solve_callback

        # Get joint group information
        joint_group_info = self.scene.getJointGroupInfo(self.joint_group)
        self._q_indices = joint_group_info.q_indices
        joint_names = np.array(self.scene.getJointNames())
        self._joint_names = joint_names[self._q_indices]

        # Construct the IK solver
        self._ik_solver = RoboPlanIK(
            scene=self.scene,
            group_name=self.joint_group,
            base_frame=self.base_link,
            tip_frame=self.tip_link,
            options=options,
        )

        # Set initial states based on the state of the scene
        self.last_joint_positions = self.scene.getCurrentJointPositions()
        se3_pose = scene.forwardKinematics(self.last_joint_positions, self.tip_link)
        self._current_pose = se3ToPose(se3_pose)
        self._target_pose = self._current_pose

        # Create a separate node for the marker server with its own executor. There are likely
        # better solutions but in the interim this removes all latency to let updates be processed
        # as soon as they come in, otherwise the server seems to get hung up.
        self._marker_node = Node(node_name)
        self._ik_server = InteractiveMarkerServer(self._marker_node, self.namespace)
        self._create_interactive_marker(self._current_pose)
        self._ik_server.applyChanges()

        self._marker_executor = SingleThreadedExecutor()
        self._marker_executor.add_node(self._marker_node)
        self._marker_thread = threading.Thread(
            target=self._spin_marker_executor, daemon=True
        )
        self._marker_thread.start()

        # Create the namespaced joint state publisher and message
        qos = QoSProfile(
            depth=1,
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            durability=QoSDurabilityPolicy.VOLATILE,
        )
        self._joint_state_pub = self.node.create_publisher(
            JointState, os.path.join(self.namespace, "joint_states"), qos
        )

        self._joint_state_msg = JointState()
        self._joint_state_msg.name = self.scene.getJointNames()
        self._joint_state_msg.header.frame_id = ""

        self.node.get_logger().info("Constructed interactive marker controller")
        self.node.get_logger().info(f"Joint group: {self.joint_group}")
        self.node.get_logger().info(f"Joints: {self._joint_names}")
        self.node.get_logger().info(f"Base link: {self.base_link}")
        self.node.get_logger().info(f"End-effector: {self.tip_link}")

    def _spin_marker_executor(self):
        try:
            self._marker_executor.spin()
        except ExternalShutdownException:
            pass

    def shutdown(self):
        self._marker_executor.shutdown()
        self._marker_thread.join(timeout=0.25)
        self._marker_node.destroy_node()

    def _create_interactive_marker(self, pose: Pose):
        """
        Constructs a new interactive marker at the specified pose.
        """
        int_marker = InteractiveMarker()
        int_marker.header.frame_id = self.base_link
        int_marker.name = "ik_target"
        int_marker.description = f"IK Target Pose for {self.joint_group}"
        int_marker.pose = pose
        int_marker.scale = 0.2

        sphere_marker = Marker()
        sphere_marker.type = Marker.SPHERE
        sphere_marker.scale.x = 0.025
        sphere_marker.scale.y = 0.025
        sphere_marker.scale.z = 0.025
        sphere_marker.color.r = 0.0
        sphere_marker.color.g = 0.5
        sphere_marker.color.b = 1.0
        sphere_marker.color.a = 1.0

        sphere_control = InteractiveMarkerControl()
        sphere_control.always_visible = True
        int_marker.controls.append(sphere_control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 1.0
        control.orientation.y = 0.0
        control.orientation.z = 0.0
        control.name = "move_x"
        control.interaction_mode = InteractiveMarkerControl.MOVE_AXIS
        int_marker.controls.append(control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 0.0
        control.orientation.y = 1.0
        control.orientation.z = 0.0
        control.name = "move_y"
        control.interaction_mode = InteractiveMarkerControl.MOVE_AXIS
        int_marker.controls.append(control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 0.0
        control.orientation.y = 0.0
        control.orientation.z = 1.0
        control.name = "move_z"
        control.interaction_mode = InteractiveMarkerControl.MOVE_AXIS
        int_marker.controls.append(control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 1.0
        control.orientation.y = 0.0
        control.orientation.z = 0.0
        control.name = "rotate_x"
        control.interaction_mode = InteractiveMarkerControl.ROTATE_AXIS
        int_marker.controls.append(control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 0.0
        control.orientation.y = 1.0
        control.orientation.z = 0.0
        control.name = "rotate_y"
        control.interaction_mode = InteractiveMarkerControl.ROTATE_AXIS
        int_marker.controls.append(control)

        control = InteractiveMarkerControl()
        control.orientation.w = 1.0
        control.orientation.x = 0.0
        control.orientation.y = 0.0
        control.orientation.z = 1.0
        control.name = "rotate_z"
        control.interaction_mode = InteractiveMarkerControl.ROTATE_AXIS
        int_marker.controls.append(control)

        self._ik_server.insert(
            int_marker, feedback_callback=self._marker_feedback_callback
        )

    def _marker_feedback_callback(self, feedback: InteractiveMarkerFeedback):
        """
        Callback function to process pose updates from the iMarker.

        Args:
            feedback: InteractiveMarkerFeedback message
        """
        # Record the iMarker state as it is updated and solve
        if feedback.event_type == InteractiveMarkerFeedback.POSE_UPDATE:
            self._target_pose = feedback.pose
            self.solve_ik()

    def solve_ik(self):
        """
        Solve IK for the latest set target pose.
        """
        # Get the transform for the target pose and solve
        transform = poseToSE3(self._target_pose)
        joint_positions = self._ik_solver.solve_ik(
            transform, seed_state=self.last_joint_positions
        )

        if joint_positions is not None:
            self.last_joint_positions = joint_positions
            if self.solve_callback:
                self.solve_callback(joint_positions)
        else:
            self.node.get_logger().warn("IK failed to find solution for target pose")

    def publish_joint_states(self):
        """
        Publish the latest solution's joint states.
        """
        self._joint_state_msg.header.stamp = self.node.get_clock().now().to_msg()
        self._joint_state_msg.position = self.last_joint_positions.tolist()
        self._joint_state_pub.publish(self._joint_state_msg)
